utils/property-propagation.stanza

```diff
@@ -0,0 +1,220 @@
+defpackage ocdb/property-propagation : 
+  import core
+  import collections
+  import jitx
+  import jitx/commands
+  import ocdb/property-structs
+  import ocdb/tolerance
+
+;==============================================================================
+;============================= Public API =====================================
+;==============================================================================
+; TODO: move to dedicated package, eg ocdb/introspection-utils? 
+doc: "Returns a list of unique nets within a module (nets that have a  \
+      unique set of pins associated with them)"
+public defn unique-nets (module:JITXObject) -> Seq<JITXObject> : 
+  val visited = HashSet<Tuple<Int>>()
+  for net in nets(module) filter : 
+    ; println("Checking if net (%_) is unique: %," % [net, seq(ref, connected-pins(net))])
+    val pin-hash = connected-pins(net)
+      $> seq{hash, _}
+      $> lazy-qsort
+      $> to-tuple
+    val result = not visited[pin-hash]
+    ; within indented() : 
+    ;   println(result)
+    if result :
+      add(visited, pin-hash)
+    result
```

Comment:
It seems that `connect-groups` from checks.stanza is doing what you want in a more optimal way (linear in the number of component pins and removes dnp pins before searching for connected nets). Between here and `pins-on-net` you are calling `connected-pins` potentially 100 times on the same net if the net is made up in esir with
```
pin a
for i in 0 to 100
  net (a res.p[i])
```
+ sorting is n log(n) making the algo possibly non-linear + dnp pins are included in this n

Reply:
I do not want to reinvent introspections, but go directly to the plugin.

Reply:
Here is the logic there
```
defn connect-groups (module:JITXObject) -> Connections :
  val visited-connections = HashSet<Pin>()
  val connected-groups = Vector<Tuple<Pin>>()
  val non-connected-pins = Vector<Pin>()

  for p in all-populated-pins(module) do :
    ; p is not added to visited-connections because we already looped over it, so it is clear.
    if visited-connections[p] :
      None()
    else :
      val connections = connected-pins(p)
      if empty?(connections) :
        add(non-connected-pins, p)
      else :
        add-all(visited-connections, connections)
        val connected-group = to-tuple $ cat([p], connections)
        add(connected-groups, connected-group)

  Connections(to-tuple(connected-groups),
              to-tuple(non-connected-pins))

; Returns all the pins that are electrically-meaningful, ie are not
; from components marked do-not-place
defn all-populated-pins (module:JITXObject) -> Tuple<Pin> :
  populated-components(module)
    $> seq-cat{pins, _}
    $> to-tuple ; FIXME! There can be a lot, do not to-tuple?
    $> {_ as Tuple<Pin>}

; Returns component instances not marked DNP
defn populated-components (module:JITXObject) -> Tuple<JITXObject> : 
  component-instances(module) 
    $> filter{{not dnp-component?(_)}, _}
    $> to-tuple
```

Reply:
And like the TODO mentions, this needs to be added as an introspection since it's redoing the work of flattening nets.

Reply:
Unsure what you mean? In here it seems you can make the loop body linear in the length of the net without sorting:
```
doc: "Returns a list of unique nets within a module (nets that have a  \
      unique set of pins associated with them)"
public defn unique-nets (module:JITXObject) -> Seq<JITXObject> : 
  val visited = HashSet<Pin>()
  for net in nets(module) filter : 
    val pins = connected-pins(net)
    val result = not empty?(pins) and not visited[pins[0]]
    if result :
      add-all(visited, pins)
    result
```

Reply:
If the order of the pins returned by `connected-pins` (btw: should be `pins-on-net`, that's a bug) is different the hash will be different.

Reply:
I wouldn't focus on this too much - it will be moved to the compiler to work on flattened nets (instead of using net statements)

Reply:
Yes, adding an introspection to return the connected nets directly is easy and this code will be replaced. Actually if there were to be a lot of dnps, `connected-group` from check.stanza could actually be quicker than an introspection giving out the connected nets and then filtering out dnps

Reply:
> If the order of the pins returned by connected-pins (btw: should be pins-on-net, that's a bug) is different the hash will be different.
There is no manually hashing needed